### PR TITLE
[ONNX] Update typing and error messages in symbolic_helper

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.] 

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

### CODE REVIEW

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -148,36 +186,43 @@ def _maybe_get_scalar(value):
...
-        raise RuntimeError(
-            f"ONNX symbolic expected node type `prim::TupleConstruct`, "
-            f"got `{tuple_node}`"
+        raise errors.SymbolicValueError(
+            f"ONNX symbolic expected node type 'prim::TupleConstruct', "
+            f"got '{tuple_node.kind()}'.",
+            tuple_value,
         )
-    return list(tuple_node.inputs())
+    return tuple(tuple_node.inputs())
```
Comment (AllenTiTaiWang) [inline torch/onnx/symbolic_helper.py:216]: I see this output being consumed in dequantized_helper with slicing. Is here changed to tuple a good idea?

File: torch/onnx/errors.py [hunk 2]
```diff
@@ -42,3 +50,47 @@ def __init__(
...
+
+
+class SymbolicValueError(OnnxExporterError):
+    """Errors around TorchScript values and nodes."""
+
+    def __init__(self, msg: str, value: _C.Value):
+        message = (
+            f"{msg}  [Caused by the value '{value}' (type '{value.type()}') in the "
+            f"TorchScript graph. The containing node has kind '{value.node().kind()}'.] "
+            f"\n    (node defined in {value.node().sourceRange() or 'unknown source code location'})"
```
Comment (AllenTiTaiWang) [inline torch/onnx/errors.py:62]: If the message actually showing `'unknown source code location'` is not very useful and redundant. Possibly making this nothing or specify the location?

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:216]: I was hoping to mark it as an immutable value. Confirmed offline that dequantized_helper is fine

File: torch/onnx/errors.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/errors.py:62]: Done

**Comment by justinchuby** [comment]
Closed in favor of #83007

**Pull request closed without merging**